duckplyr 1.0.0 (#724)
tidyverse/tidyverse.org: content/blog/duckplyr-1-0-0/index.Rmd

### For large data

For large data, duckplyr is a legitimate alternative to dtplyr and dbplyr.

With large datasets, you want:

- input data in an efficient format, like Parquet files, which duckplyr allows thanks to its ingestion functions like `read_parquet_duckdb()`.
- efficient computation, which duckplyr provides via DuckDB's holistic optimization, without your having to use another syntax than dplyr.
- the output to not clutter all the memory, which duckplyr supports through two features:
    - the [control of automatic materialization](https://duckplyr.tidyverse.org/articles/prudence.html) (collection of results into memory) thanks to the `prudence` parameter. You can disable automatic materialization completely or, as a compromise, disable it up to a certain output size.
    - [computation to files](https://duckplyr.tidyverse.org/reference/compute_file.html) using  `compute_parquet()` or `compute_csv()`.

A drawback of analyzing large data with duckplyr is that the limitations of duckplyr won't be compensated by fallbacks, since fallbacks to dplyr necessitate putting data into memory.

> Therefore, if your pipeline encounters fallbacks, you might want to work around them by converting the duck frame into a table through `compute()` then running SQL code through the experimental `read_sql_duckdb()` function. Again, over time, we expect more native support for dplyr functionality.

Author: @krlmlr could we tweak the example to use `ceiling()` that isn't supported I think? So it'd look more realistic. (I do not know SQL :see_no_evil: )

slug: duckplyr-1-0-0
title: duckplyr fully joins the tidyverse!
date: 2025-02-13
author: Kirill Müller and Maëlle Salmon
description: >
    duckplyr 1.0.0 is on CRAN and part of the tidyverse!
    A drop-in replacement for dplyr, powered by DuckDB for speed.
    It is the most dplyr-like of dplyr backends.

photo:
  url: https://www.pexels.com/photo/a-mallard-duck-on-water-6918877/
  author: Kiril Gruev

# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
categories: [package]
tags:
  - duckplyr
  - dplyr
  - tidyverse
---

<!--
TODO:
* [x] Look over / edit the post's title in the yaml
* [x] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [x] Find photo & update yaml metadata
* [x] Create `thumbnail-sq.jpg`; height and width should be equal
* [x] Create `thumbnail-wd.jpg`; width should be >5x height
* [x] `hugodown::use_tidy_thumbnails()`
* [x] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
-->

We're very chuffed to announce the release of [duckplyr](https://duckplyr.tidyverse.org) 1.0.0.

> duckplyr is a drop-in, fully compatible replacement for dplyr, powered by [DuckDB](https://duckdb.org/) for speed.

Reviewer: Rather than repeating the start of the first section here, I'd say something like "duckdbplyr is a new dplyr backend powered by duckdb".

It joins the rank of dplyr backends together with [dtplyr](https://dtplyr.tidyverse.org) and [dbplyr](https://dbplyr.tidyverse.org).
You can use it instead of dplyr for data small or large.

<!-- FIXME:

We have many more dplyr backends, the two above are just from the tidyverse.
GitHub search: https://github.com/search?q=org%3Acran+%2FS3method%5B%28%5D%28mutate%7Csummarise%29+*%2C%2F&type=code
Do we need an "awesome dplyr" like https://github.com/krlmlr/awesome-vctrs/?

-->

You can install it from CRAN with:

```{r, eval = FALSE}
install.packages("duckplyr")
```

In this article, we'll show how duckplyr can help you with data of different size, explain how you can help improve the package, and ... .

## A drop-in replacement for dplyr

> The duckplyr package is a _drop-in replacement for dplyr_ that uses _DuckDB for speed_.

Reviewer: I think you need a sentence or two here as to why it's needed compared to dbplyr.

Reviewer 2: Done below: https://github.com/tidyverse/tidyverse.org/pull/724/files#diff-30be2863e4e092b08c7c1ad595164bec2f424a02f175c56e32f85d3475559a94R102-R104

> Like with other dplyr backends like dtplyr and dbplyr, duckplyr allows you to get faster results without learning a different syntax.
> Unlike other dplyr backends, duckplyr does not require you to change existing code or learn specific idiosyncracies.
> Not only is the syntax the same, the semantics are too!

You can simply _drop_ duckplyr into your pipeline by loading it, then computations will be efficiently carried out by DuckDB.
DuckDB is a fast in-memory analytical database system[^duckdb].

[^duckdb]: If you haven't heard about it, you can watch [Hannes Mühleisen's keynote at posit::conf(2024)](https://www.youtube.com/watch?v=GELhdezYmP0&feature=youtu.be).

```{r}
library(conflicted)
library(duckplyr)
conflict_prefer("filter", "dplyr", quiet = TRUE)
library(babynames)

out <- babynames |>
  mutate(prevalence = if_else(prop >= 0.01, "frequent", "rare")) |>
  summarize(
    .by = c(sex, year, prevalence),
    babies_n = sum(n)
  ) |>
  filter(sex == "F")
class(out)
out
```

Like with other dplyr backends like dtplyr and dbplyr, duckplyr allows you to get faster results.

> Unlike other dplyr backends, duckplyr does not require you to learn a different syntax.

Reviewer: What backend requires that you learn a different syntax?

Author: Oops, I stand corrected!

The duckplyr package is fully compatible with dplyr: if an operation cannot be carried out with DuckDB, it is automatically outsourced to dplyr.
Over time, we expect fewer and fewer fallbacks to dplyr to be needed.

## How to use duckplyr

To _replace_ dplyr with duckplyr, you can:

- Load duckplyr and then keep your pipeline as is. Calling `library(duckplyr)` overwrites dplyr methods, enabling duckplyr for the entire session no matter how data.frames are created.
    This is shown in the example above.

- Create individual "duck frames" using _conversion functions_ like `duckdb_tibble()` or `as_duckdb_tibble()`, or _ingestion functions_ like `read_csv_duckdb()`.

Then, the data manipulation pipeline uses the exact same syntax as a dplyr pipeline.
The duckplyr package performs the computation using DuckDB.

```{r}
# Undo the effect of library(duckplyr)

> methods_restore()

Reviewer: Could we drop this for the purposes of simplicity?

Author: Hidden now

Reviewer 2: In my view, hiding is worse than doing without. Now, with the benchmark, we cannot do without. I propose to show it and talk about it.

Author: Added it back.

out <- babynames |>
  as_duckdb_tibble() |>
  mutate(prevalence = if_else(prop >= 0.01, "frequent", "rare")) |>
  summarize(
    .by = c(sex, year, prevalence),
    babies_n = sum(n)
  ) |>
  filter(sex == "F")
class(out)
```

In both cases, printing the result only shows the first few rows, as with dbplyr.

```{r}
out
```

The result can finally be materialized to memory, or computed temporarily, or computed to a file.

```{r}
# to memory
nrow(out)
# or for instance collect(out)

# to a file
csv_file <- withr::local_tempfile()
compute_csv(out, csv_file)
fs::file_size(csv_file)
```

When duckplyr itself does not support specific functionality, it falls back to dplyr.
For instance, filtering on grouped data is not supported yet, still it works thanks to the fallback mechanism.

```{r}
babynames |>

>   filter(n > 10000, .by = "name")

Reviewer: Ideally this would use an expression where grouping matters (e.g. `filter(n == max(n), .by = c(year, sex))`.

```

For performance reasons, the output order of the result is not guaranteed to be stable.
If you need a stable order, you can use `arrange()` or force output order stability by setting an environment variable.
This and other limitations are documented in [`vignette("limits")`](https://duckplyr.tidyverse.org/articles/limits.html).

## Large data

> For large data, duckplyr is a legitimate alternative to dtplyr and dbplyr.

Reviewer: I think you need to be explicit here that large = larger than memory.

With large datasets, you want:

- input data in an efficient format, like Parquet files, which duckplyr allows thanks to its ingestion functions like `read_parquet_duckdb()`.
- efficient computation, which duckplyr provides via DuckDB's holistic optimization, without your having to use another syntax than dplyr.
- the output to not clutter all the memory, which duckplyr supports through two features:
    - computation to files using  [`compute_parquet()`](https://duckplyr.tidyverse.org/reference/compute_file.html) or [`compute_csv()`](https://duckplyr.tidyverse.org/reference/compute_file.html).
    - the control of automatic materialization (collection of results into memory). You can disable automatic materialization completely or, as a compromise, disable it up to a certain output size. See [`vignette("prudence")`](https://duckplyr.tidyverse.org/articles/prudence.html) for details.

See [`vignette("large")`](https://duckplyr.tidyverse.org/articles/large.html) for a walkthrough and more details.

A drawback of analyzing large data with duckplyr is that the limitations of duckplyr won't be compensated by fallbacks, since fallbacks to dplyr necessitate putting data into memory.

> Therefore, if your pipeline encounters fallbacks, you might want to work around them by converting the duck frame into a table through `compute()` then running SQL code through the experimental `read_sql_duckdb()` function.

Reviewer: I think it would be simpler to recommend running code to filter/sample/summarise to smaller than memory, then leave this advice to vignette.

Author: I'm removing this entirely for now.

---
output: hugodown::hugo_document

slug: duckplyr-1-0-0
title: duckplyr fully joins the tidyverse!
date: 2025-02-13
author: Kirill Müller and Maëlle Salmon
description: >

>     duckplyr 1.0.0 is on CRAN and part of the tidyverse!

Reviewer 2: Working on that part, duckdb 1.2.0 is about to be released tomorrow.

  url: https://www.pexels.com/photo/a-mallard-duck-on-water-6918877/
  author: Kiril Gruev

# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
categories: [package]
tags:
  - duckplyr
  - dplyr
  - tidyverse
---

<!--
TODO:
* [x] Look over / edit the post's title in the yaml
* [x] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [x] Find photo & update yaml metadata
* [x] Create `thumbnail-sq.jpg`; height and width should be equal
* [x] Create `thumbnail-wd.jpg`; width should be >5x height
* [x] `hugodown::use_tidy_thumbnails()`
* [x] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
-->

We're very chuffed to announce the release of [duckplyr](https://duckplyr.tidyverse.org) 1.0.0.
This is a new dplyr backend powered by [DuckDB](https://duckdb.org/), a fast in-memory analytical database system[^duckdb].
It joins the rank of dplyr backends together with [dtplyr](https://dtplyr.tidyverse.org) and [dbplyr](https://dbplyr.tidyverse.org).
You can use it instead of dplyr for data small or large.

<!-- FIXME:

> We have many more dplyr backends, the two above are just from the tidyverse.

Author: is the action needed here to create this repo? under the cynkra org?

GitHub search: https://github.com/search?q=org%3Acran+%2FS3method%5B%28%5D%28mutate%7Csummarise%29+*%2C%2F&type=code
Do we need an "awesome dplyr" like https://github.com/krlmlr/awesome-vctrs/?

-->

You can install it from CRAN with:

```{r, eval = FALSE}

> install.packages("duckplyr")

Reviewer 2: With https://github.com/tidyverse/tidyverse/pull/346, we can also `install.packages("tidyverse")` .

Author: Right, but won't the post be published before the PR is merged and the tidyverse package is released on CRAN?

```

> In this article, we'll show how duckplyr can help you with data of different size, and explain how you can help improve the package.

Reviewer 2: This no longer plays well with the previous sentence "You can use it instead of dplyr for data small or large". Perhaps combine?

```suggestion
This article shows how duckplyr can be used instead of dplyr with data of different size, explain how you can help improve the package, and share a selection of other resources.
```

## A drop-in replacement for dplyr

Imagine you have to wrangle a huge dataset.
Here we generate one using the [data generator from the TPC-H benchmark](https://duckdb.org/2024/04/02/duckplyr.html#benchmark-tpc-h-q1).

```{r}
lineitem <- duckdb:::sql("INSTALL tpch; LOAD tpch; CALL dbgen(sf=1); FROM lineitem;")
lineitem <- tibble::as_tibble(lineitem)
dplyr::glimpse(lineitem)
```

We could transform the data using dplyr but we could also transform it using a tool that'll scale well to ever larger data: duckplyr.
The duckplyr package is a _drop-in replacement for dplyr_ that uses _DuckDB for speed_.
You can simply _drop_ duckplyr into your pipeline by loading it, then computations will be efficiently carried out by DuckDB.

[^duckdb]: If you haven't heard about it, you can watch [Hannes Mühleisen's keynote at posit::conf(2024)](https://www.youtube.com/watch?v=GELhdezYmP0&feature=youtu.be).

Below, we express the standard "TPC-H benchmark query 1" in dplyr syntax, but execute it with duckplyr.

```{r}

> library(dplyr)
> library(conflicted)
> library(duckplyr)

Reviewer 2: ```suggestion
library(conflicted)
library(duckplyr)
```

This should remove the nasty output. You could also `library(dplyr)` before `library(duckplyr)` for extra verbosity.

Then, the data manipulation pipeline uses the exact same syntax as a dplyr pipeline.
The duckplyr package performs the computation using DuckDB.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Undo the effect of library(duckplyr)
methods_restore()
```

```{r}
out <- lineitem |>
  duckplyr::as_duckdb_tibble() |> # this is the only change :-)
  select(l_shipdate, l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax) |>
  filter(l_shipdate <= !!as.Date("1998-09-02")) |>
  select(l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax) |>
  summarise(
    sum_qty = sum(l_quantity),
    sum_base_price = sum(l_extendedprice),
    sum_disc_price = sum(l_extendedprice * (1 - l_discount)),
    sum_charge = sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)),
    avg_qty = mean(l_quantity),
    avg_price = mean(l_extendedprice),
    avg_disc = mean(l_discount),
    count_order = n(),
    .by = c(l_returnflag, l_linestatus)
  ) |>
  arrange(l_returnflag, l_linestatus)
```

The result can finally be materialized to memory, or computed temporarily, or computed to a file.

```{r}
# to memory
nrow(out)

> # or for instance collect(out)

Reviewer 2: How and why does `nrow()` materialize to memory?

I had this in a separate paragraph in my edits.

# to a file
csv_file <- withr::local_tempfile()
compute_csv(out, csv_file)
fs::file_size(csv_file)
```

Operations not yet supported by duckplyr are automatically outsourced to dplyr.
For instance, filtering on grouped data is not supported yet, still it works thanks to the fallback mechanism.

```{r}

> # filter(n == max(n), .by = c(year, sex)

Author: no, I entirely removed the babynames example in favor of lineitems.

lineitem |>
  duckplyr::as_duckdb_tibble() |>
  select(l_shipdate, l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax) |>

>   filter(l_quantity == max(l_quantity), .by = c(l_returnflag, l_linestatus))

Reviewer 2: I wonder why you're seeing the message here in the rendered output. Have you set an environment variable?

Author: I'll knit again once the new version is on CRAN, this way the message won't have the typo (https://github.com/tidyverse/duckplyr/pull/611)

```

> For performance reasons, the output order of the result is not guaranteed to be stable.
> If you need a stable order, you can use `arrange()` or force output order stability by setting an environment variable.
> This and other limitations are documented in [`vignette("limits")`](https://duckplyr.tidyverse.org/articles/limits.html).
> 

Reviewer 2: This now has little value because our TPC-H example has a final `arrange()`, and nobody would be able to tell from our new example unless we show them.

```suggestion
```

    ) |>
    arrange(l_returnflag, l_linestatus) 
}
```

Here is the function that runs it with duckplyr. 
The only differences are the lines `duckplyr::as_duckdb_tibble()` and `collect()` (for ensuring materialization otherwise the comparison isn't fair).

```{r}
tpch_duckplyr <- function(lineitem) {
  lineitem |>
    duckplyr::as_duckdb_tibble() |> # difference 1/2
    select(l_shipdate, l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax) |>
    filter(l_shipdate <= !!as.Date("1998-09-02")) |>
    select(l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax) |>
    summarise(
      sum_qty = sum(l_quantity),
      sum_base_price = sum(l_extendedprice),
      sum_disc_price = sum(l_extendedprice * (1 - l_discount)),
      sum_charge = sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)),
      avg_qty = mean(l_quantity),
      avg_price = mean(l_extendedprice),
      avg_disc = mean(l_discount),
      count_order = n(),
      .by = c(l_returnflag, l_linestatus)
    ) |>
    arrange(l_returnflag, l_linestatus) |>
    collect() # difference 2/2
}
```

And now we compare the two:

```{r}
bench::mark(
  tpch_dplyr(lineitem),

>   tpch_duckplyr(lineitem),

Author: As you prefer but I find the nested calls less easy to read?

>   tpch_duckplyr(lineitem),

Author: especially as some readers might not be familiar with `bench::mark()`, so the chunk might look difficult with the nested call?
At the same time I understand that not introducing a new function would save space.

  check = ~ all.equal(.x, .y, tolerance = 1e-10)
)
```

In this benchmark, the pipeline run with duckplyr is clearly faster than the pipeline run with dplyr.
Start using duckplyr today by attaching it and running your existing dplyr code.
Many operations will be carried out with DuckDB, faster than with dplyr.

## Data larger than memory

For data larger than memory, duckplyr is a legitimate alternative to dtplyr and dbplyr.

With datasets that approach or surpass the size of your machine's RAM, you want:

- input data in an efficient format, like Parquet files, which duckplyr allows thanks to its ingestion functions like `read_parquet_duckdb()`.
- efficient computation, which duckplyr provides via DuckDB's holistic optimization, without your having to adapt your code.
- the output to not clutter all the memory, which duckplyr supports through two features:
    - computation to files using  [`compute_parquet()`](https://duckplyr.tidyverse.org/reference/compute_file.html) or [`compute_csv()`](https://duckplyr.tidyverse.org/reference/compute_file.html).

>     - the control of automatic materialization (collection of results into memory). You can disable automatic materialization completely or, as a compromise, disable it up to a certain output size. See [`vignette("prudence")`](https://duckplyr.tidyverse.org/articles/prudence.html) for details.

Author: I find it less clear as these two things are related to outputs.

>     - the control of automatic materialization (collection of results into memory). You can disable automatic materialization completely or, as a compromise, disable it up to a certain output size. See [`vignette("prudence")`](https://duckplyr.tidyverse.org/articles/prudence.html) for details.

Author: In your edits, you had removed the notion of controlling automatic materialization, why? I found it important in this section.

```markdown
- large results written to files using  [`compute_parquet()`](https://duckplyr.tidyverse.org/reference/compute_file.html) or [`compute_csv()`](https://duckplyr.tidyverse.org/reference/compute_file.html).
- small results processed seamlessly with dplyr, using all verbs and functions.
```

Special thanks to Joe Thorley ([&#x0040;joethorley](https://github.com/joethorley)) for help with choosing the right words.

Eager to learn more about duckplyr -- beside by trying it out yourself?
The pkgdown website of duckplyr features several [articles](https://duckplyr.tidyverse.org/articles/).

> Furthermore, the blog post ["duckplyr: dplyr Powered by DuckDB"](https://duckdb.org/2024/04/02/duckplyr.html) by Hannes Mühleisen provides some context on duckplyr including its inner workings, as also seen in a [section](https://blog.r-hub.io/2025/02/13/lazy-meanings/#duckplyr-lazy-evaluation-and-prudence) of the R-hub blog post ["Lazy introduction to laziness in R"](https://blog.r-hub.io/2025/02/13/lazy-meanings/) by Maëlle Salmon, Athanasia Mo Mowinckel and Hannah Frick. 

Reviewer 2: Move to a separate section before the acknowledgments?

Below, we express the standard "TPC-H benchmark query 1" in dplyr syntax, but execute it with duckplyr.

```{r}
library(conflicted)
library(duckplyr)
conflict_prefer("filter", "dplyr", quiet = TRUE)
out <- lineitem |>
  select(l_shipdate, l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax) |>
  filter(l_shipdate <= !!as.Date("1998-09-02")) |>
  select(l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax) |>
  summarise(
    sum_qty = sum(l_quantity),
    sum_base_price = sum(l_extendedprice),
    sum_disc_price = sum(l_extendedprice * (1 - l_discount)),
    sum_charge = sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)),
    avg_qty = mean(l_quantity),
    avg_price = mean(l_extendedprice),
    avg_disc = mean(l_discount),
    count_order = n(),
    .by = c(l_returnflag, l_linestatus)
  ) |>
  arrange(l_returnflag, l_linestatus)
out
```

Like with other dplyr backends like dtplyr and dbplyr, duckplyr allows you to get faster results without learning a different syntax.

> Unlike other dplyr backends, duckplyr does not require you to change existing code or learn specific idiosyncracies.

Author: I'm wondering whether this sentence is slightly misleading: to use duckplyr efficiently, it's probably best to read the "limits" vignette, which is akin to learning about idiosyncrasies?
